# Université Paul Sabatier

M1IAFA - Recherche d'information

**TP 5**

Enseignants : Lynda Tamine et Jesús Lovón

Notebook proposé par : José G. Moreno


2024

---

💡 Penser à développer des scripts et fonctions auxiliaires qui vont permettre réutiliser les commandes récurrentes dans ce TP et les prochains. Ceci aussi vous permettra de garder de bon pratique du code et faciliter vos débogages.

---


### Attention ❗❗ Pour la note du TP :
🚨 *Questions de code* : Remplir le code manquant dans les parties correspondantes (le code commenté bénéficie des meilleures notes).

🚨 *Questions ouvertes* : Ecrivez votre réponse textuelle sous forme de commentaires dans les cellules correspondantes.

🚨 *Laissez vos sorties* pour les cellules où vous écrivez le code. Les sorties vides (notebook ou cellules non exécutées) correspondent à 0 points.

---

# TP 5. PyTerrier - Neural Re-Ranking

Dans ce TP vous allez :

 - reclasser des documents en utilisant des modèles neuronaux comme KNRM, Vanilla BERT, EPIC et monoT5.


# 0. Installation


In [1]:
!pip install -q --upgrade python-terrier

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 107.6/107.6 kB 3.4 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.5/1.5 MB 13.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 69.6/69.6 kB 8.2 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 335.2/335.2 kB 16.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 48.8/48.8 kB 5.9 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 15.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 287.4/287.4 kB 16.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.1/45.1 kB 6.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.4/45.4 kB 6.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 23.1 MB/s eta 0:00:00
  Preparing metadata (setup

# Installation des plugins Pyterrier  

Nous installons les plugins PyTerrier [OpenNIR](https://opennir.net/) et [monoT5](https://github.com/terrierteam/pyterrier_t5). Vous pouvez ignorer sans risque les erreurs de version des paquets.

In [2]:
!pip install -q --upgrade git+https://github.com/Georgetown-IR-Lab/OpenNIR
!pip install -q --upgrade git+https://github.com/terrierteam/pyterrier_t5

  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 114.3/114.3 kB 1.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 158.1/158.1 kB 5.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 235.5/235.5 kB 22.4 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 85.1/85.1 kB 9.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 139.3/139.3 kB 17.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.2/12.2 MB 100.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 82.2/82.2 kB 12.1 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done


# 1. Preliminary steps

In [3]:
import os
import tensorflow as tf

**[PyTerrier](https://github.com/terrier-org/pyterrier) initialization**

In [4]:
import pyterrier as pt
if not pt.started():
    pt.init()

cord19 = pt.datasets.get_dataset('irds:cord19/trec-covid')


from pyterrier.measures import * # allow for natural measure names
import onir_pt

terrier-assemblies 5.8 jar-with-dependencies not found, downloading to /root/.pyterrier...
Done
terrier-python-helper 0.0.8 jar not found, downloading to /root/.pyterrier...
Done


PyTerrier 0.10.0 has loaded Terrier 5.8 (built by craigm on 2023-11-01 18:05) and terrier-helper 0.0.8



Better speed can be achieved with apex installed from https://www.github.com/nvidia/apex.


# 2. Indexation

In [5]:
!rm -rf ./terrier_cord19/*

In [6]:
import os

pt_index_path = './terrier_cord19'

if not os.path.exists(pt_index_path + "/data.properties"):
    # create the index, using the IterDictIndexer indexer
    indexer = pt.index.IterDictIndexer(pt_index_path, blocks=True)

    # we give the dataset get_corpus_iter() directly to the indexer
    # while specifying the fields to index and the metadata to record
    index_ref = indexer.index(cord19.get_corpus_iter(),
                              fields=('abstract',),
                              meta=('docno',))

else:
    # if you already have the index, use it.
    index_ref = pt.IndexRef.of(pt_index_path + "/data.properties")

[INFO] [starting] building docstore
[INFO] If you have a local copy of https://ai2-semanticscholar-cord-19.s3-us-west-2.amazonaws.com/2020-07-16/metadata.csv, you can symlink it here to avoid downloading it again: /root/.ir_datasets/downloads/80d664e496b8b7e50a39c6f6bb92e0ef
[INFO] [starting] https://ai2-semanticscholar-cord-19.s3-us-west-2.amazonaws.com/2020-07-16/metadata.csv
docs_iter:   0%|                                    | 0/192509 s<?, ?doc/s]
https://ai2-semanticscholar-cord-19.s3-us-west-2.amazonaws.com/2020-07-16/metadata.csv: 0.0%| 0.00/269M s<?, ?B/s]
https://ai2-semanticscholar-cord-19.s3-us-west-2.amazonaws.com/2020-07-16/metadata.csv: 0.1%| 147k/269M s<03:29, 1.29MB/s]
https://ai2-semanticscholar-cord-19.s3-us-west-2.amazonaws.com/2020-07-16/metadata.csv: 0.5%| 1.44M/269M s<44.84s, 5.97MB/s]
https://ai2-semanticscholar-cord-19.s3-us-west-2.amazonaws.com/2020-07-16/metadata.csv: 2.8%| 7.43M/269M s<13.03s, 20.1MB/s]
https://ai2-semanticscholar-cord-19.s3-us-west-2.amazon

cord19/trec-covid documents:   0%|          | 0/192509 s<?, ?it/s]

<ipython-input-6-47146416b637>:11: DeprecationWarning: specifying meta and meta_lengths in IterDictIndexer.index() is deprecated, use constructor instead
  index_ref = indexer.index(cord19.get_corpus_iter(),


21:04:00.302 [ForkJoinPool-1-worker-3] WARN org.terrier.structures.indexing.Indexer - Adding an empty document to the index (6iu1dtyl) - further warnings are suppressed
21:05:15.546 [ForkJoinPool-1-worker-3] ERROR org.terrier.structures.indexing.Indexer - Could not finish MetaIndexBuilder: 
java.io.IOException: Key 8lqzfj2e is not unique: 37597,11755
For MetaIndex, to suppress, set metaindex.compressed.reverse.allow.duplicates=true
	at org.terrier.structures.collections.FSOrderedMapFile$MultiFSOMapWriter.mergeTwo(FSOrderedMapFile.java:1374)
	at org.terrier.structures.collections.FSOrderedMapFile$MultiFSOMapWriter.close(FSOrderedMapFile.java:1308)
	at org.terrier.structures.indexing.BaseMetaIndexBuilder.close(BaseMetaIndexBuilder.java:321)
	at org.terrier.structures.indexing.classical.BasicIndexer.indexDocuments(BasicIndexer.java:270)
	at org.terrier.structures.indexing.classical.BasicIndexer.createDirectIndex(BasicIndexer.java:388)
	at org.terrier.structures.indexing.Indexer.index(Inde

# 3. Re-Rankers

Commençons à explorer quelques méthodes neuronales de re-classement ! Nous pouvons les construire à partir de zéro en utilisant `onir_pt.reranker`.

Le modèle de re-ranking d'OpenNIR est composé de :
 - `ranker` (par exemple, `drmm`, `knrm`, ou `pacrr`). Ceci définit l'architecture neuronale de classement.
 - `vocab` (par exemple, `wordvec_hash`, ou `bert`). Ceci définit comment le texte est encodé par le modèle. Cette approche rend facile l'échange de différentes représentations de texte.

L'exécution de cette ligne prendra quelques minutes car elle télécharge et prépare les vecteurs de mots.

In [7]:
knrm = onir_pt.reranker('knrm', 'wordvec_hash', text_field='abstract')

config file not found: config
[2024-04-20 21:05:39,397][WordvecHashVocab][DEBUG] [starting] downloading https://dl.fbaipublicfiles.com/fasttext/vectors-english/wiki-news-300d-1M.vec.zip


[2024-04-20 21:05:46,217][onir.util.download][WARNING] no hash provided for https://dl.fbaipublicfiles.com/fasttext/vectors-english/wiki-news-300d-1M.vec.zip; consider adding expected_md5="3cc8839ac3fa9a6187149b1e73328b2a" to ensure data integrity.
[2024-04-20 21:05:46,225][onir.util.download][DEBUG] downloaded https://dl.fbaipublicfiles.com/fasttext/vectors-english/wiki-news-300d-1M.vec.zip [6.61s] [682M] [134MB/s]
[2024-04-20 21:05:46,233][WordvecHashVocab][DEBUG] [finished] downloading https://dl.fbaipublicfiles.com/fasttext/vectors-english/wiki-news-300d-1M.vec.zip [6.84s]
[2024-04-20 21:05:46,234][WordvecHashVocab][DEBUG] [starting] extracting vecs
[2024-04-20 21:06:05,356][WordvecHashVocab][DEBUG] [finished] extracting vecs [19.12s]
[2024-04-20 21:06:05,356][WordvecHashVocab][DEBUG] [starting] loading vecs into memory
[2024-04-20 21:08:25,285][WordvecHashVocab][DEBUG] [finished] loading vecs into memory [02:20]
[2024-04-20 21:08:25,637][WordvecHashVocab][DEBUG] [starting] writing

Voyons comment ces modèles fonctionnent pour le classement !

In [8]:
tfidf = pt.BatchRetrieve(index_ref, wmodel="TF_IDF") % 50
get_text = pt.text.get_text(cord19, 'abstract') #>> pt.apply.title_abstract(lambda r: r['title'] + ' ' + r['abstract'])


In [9]:
topics = cord19.get_topics(variant='description')
qrels = cord19.get_qrels()

[INFO] [starting] https://ir.nist.gov/covidSubmit/data/topics-rnd5.xml
[INFO] [finished] https://ir.nist.gov/covidSubmit/data/topics-rnd5.xml: s] [18.7kB] [41.6MB/s]
[INFO] [starting] https://ir.nist.gov/covidSubmit/data/qrels-covid_d5_j0.5-5.txt
[INFO] [finished] https://ir.nist.gov/covidSubmit/data/qrels-covid_d5_j0.5-5.txt: s] [1.14MB] [3.61MB/s]


In [10]:
SEED=42

from sklearn.model_selection import train_test_split

tr_va_topics, test_topics = train_test_split(topics, test_size=15, random_state=SEED)
train_topics, valid_topics =  train_test_split(tr_va_topics, test_size=5, random_state=SEED)


test_qrels = qrels # seulement les annotations des topics en réponse sont utilisés, donc pas de problème si on utilise tout
train_qrels = qrels
valid_qrels = qrels

In [11]:

# build a sub-pipeline to get the concatenated title and abstract text
pipeline = tfidf >> get_text >> knrm
pt.Experiment(
    [tfidf, pipeline],
    test_topics,
    qrels,
    names=['TFIDF', 'TFIDF >> KNRM'],
    eval_metrics=[AP(rel=2), nDCG, nDCG@10, P(rel=2)@10]
)

[2024-04-20 21:09:01,366][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:09:01,938][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/188 s<?, ?it/s]

[2024-04-20 21:09:03,518][onir_pt][DEBUG] [finished] batches: [1.58s] [188it] [119.04it/s]


name  AP(rel=2)      nDCG   nDCG@10  P(rel=2)@10
0          TFIDF   0.054568  0.123589  0.595818     0.546667
1  TFIDF >> KNRM   0.042539  0.112699  0.423870     0.360000

Cela ne fonctionne pas très bien car le modèle n'est pas entraîné ; il utilise des poids aléatoires pour combiner les scores de la matrice de similarité.

# 4. Entraînement du re-ranker

Vous pouvez entraîner des modèles de re-classement dans PyTerrier en utilisant la méthode `fit`.

In [12]:
pipeline.fit(
    train_topics,
    train_qrels,
    valid_topics,
    valid_qrels)

[2024-04-20 21:09:11,874][onir_pt][DEBUG] [starting] validation
[2024-04-20 21:09:11,875][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:09:11,876][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2024-04-20 21:09:12,057][onir_pt][DEBUG] [finished] batches: s] [63it] [348.38it/s]
[2024-04-20 21:09:12,185][onir_pt][DEBUG] [finished] validation s]
[2024-04-20 21:09:12,186][onir_pt][INFO] pre-validation: 0.0053
[2024-04-20 21:09:12,193][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:09:12,194][onir_pt][DEBUG] [starting] training
[2024-04-20 21:09:12,195][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2024-04-20 21:09:12,722][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:13,258][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:14,365][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:15,623][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:16,139][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:17,375][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:18,753][onir_pt][DEBUG] [finished] train pairs: [6.56s] [1024it] [156.16it/s]
[2024-04-20 21:09:18,758][onir_pt][DEBUG] [finished] training [6.56s]
[2024-04-20 21:09:18,759][onir_pt][INFO] training   it=0 loss=0.2451
[2024-04-20 21:09:18,759][onir_pt][DEBUG] [starting] validation
[2024-04-20 21:09:18,760][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:09:18,761][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2024-04-20 21:09:18,946][onir_pt][DEBUG] [finished] batches: s] [63it] [343.54it/s]
[2024-04-20 21:09:19,077][onir_pt][DEBUG] [finished] validation s]
[2024-04-20 21:09:19,079][onir_pt][INFO] validation it=0 map=0.0058 ndcg=0.0151 P_10=0.0700 <--
[2024-04-20 21:09:19,079][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:09:19,081][onir_pt][DEBUG] [starting] training
[2024-04-20 21:09:19,081][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2024-04-20 21:09:19,366][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:19,463][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:21,303][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:22,346][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:22,647][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:23,063][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:23,195][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:23,706][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:24,253][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:24,466][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:24,634][onir_pt][DEBUG] [finished] train pairs: [5.55s] [1024it] [184.44it/s]
[2024-04-20 21:09:24,638][onir_pt][DEBUG] [finished] training [5.56s]
[2024-04-20 21:09:24,639][onir_pt][INFO] training   it=1 loss=0.2213
[2024-04-20 21:09:24,639][onir_pt][DEBUG] [starting] validation
[2024-04-20 21:09:24,640][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:09:24,641][onir_pt][DEBUG] [starting] 

batches:   0%|          | 0/63 s<?, ?it/s]

[2024-04-20 21:09:24,861][onir_pt][DEBUG] [finished] batches: s] [63it] [287.78it/s]
[2024-04-20 21:09:24,984][onir_pt][DEBUG] [finished] validation s]
[2024-04-20 21:09:24,986][onir_pt][INFO] validation it=1 map=0.0059 ndcg=0.0152 P_10=0.0700 <--
[2024-04-20 21:09:24,986][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:09:24,987][onir_pt][DEBUG] [starting] training
[2024-04-20 21:09:24,988][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2024-04-20 21:09:25,581][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:26,024][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:26,617][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:26,793][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:27,017][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:27,368][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:27,567][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:28,999][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:30,679][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:30,899][onir_pt][DEBUG] [finished] train pairs: [5.91s] [1024it] [173.26it/s]
[2024-04-20 21:09:30,901][onir_pt][DEBUG] [finished] training [5.91s]
[2024-04-20 21:09:30,901][onir_pt][INFO] training   it=2 loss=0.2325
[2024-04-20 21:09:30,901][onir_pt][DEBUG] [starting] validation
[2024-04-20 21:09:30,901][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:09:30,904][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2024-04-20 21:09:31,185][onir_pt][DEBUG] [finished] batches: s] [63it] [224.59it/s]
[2024-04-20 21:09:31,419][onir_pt][DEBUG] [finished] validation s]
[2024-04-20 21:09:31,422][onir_pt][INFO] validation it=2 map=0.0059 ndcg=0.0152 P_10=0.0700 <--
[2024-04-20 21:09:31,422][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:09:31,423][onir_pt][DEBUG] [starting] training
[2024-04-20 21:09:31,424][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2024-04-20 21:09:31,472][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:33,526][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:33,936][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:34,394][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:34,755][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:35,019][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:36,260][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:36,928][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:37,471][onir_pt][DEBUG] [finished] train pairs: [6.05s] [1024it] [169.37it/s]
[2024-04-20 21:09:37,475][onir_pt][DEBUG] [finished] training [6.05s]
[2024-04-20 21:09:37,476][onir_pt][INFO] training   it=3 loss=0.2365
[2024-04-20 21:09:37,477][onir_pt][DEBUG] [starting] validation
[2024-04-20 21:09:37,477][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:09:37,479][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2024-04-20 21:09:37,672][onir_pt][DEBUG] [finished] batches: s] [63it] [327.43it/s]
[2024-04-20 21:09:37,794][onir_pt][DEBUG] [finished] validation s]
[2024-04-20 21:09:37,796][onir_pt][INFO] validation it=3 map=0.0061 ndcg=0.0157 P_10=0.0800 <--
[2024-04-20 21:09:37,797][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:09:37,798][onir_pt][DEBUG] [starting] training
[2024-04-20 21:09:37,798][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2024-04-20 21:09:38,179][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:38,606][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:38,809][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:39,930][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:40,505][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:40,964][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:42,810][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:43,011][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:43,568][onir_pt][DEBUG] [finished] train pairs: [5.77s] [1024it] [177.50it/s]
[2024-04-20 21:09:43,572][onir_pt][DEBUG] [finished] training [5.77s]
[2024-04-20 21:09:43,573][onir_pt][INFO] training   it=4 loss=0.2294
[2024-04-20 21:09:43,574][onir_pt][DEBUG] [starting] validation
[2024-04-20 21:09:43,575][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:09:43,578][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2024-04-20 21:09:43,862][onir_pt][DEBUG] [finished] batches: s] [63it] [222.69it/s]
[2024-04-20 21:09:44,098][onir_pt][DEBUG] [finished] validation s]
[2024-04-20 21:09:44,099][onir_pt][INFO] validation it=4 map=0.0061 ndcg=0.0157 P_10=0.0760
[2024-04-20 21:09:44,100][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:09:44,101][onir_pt][DEBUG] [starting] training
[2024-04-20 21:09:44,102][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2024-04-20 21:09:44,248][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:44,732][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:45,290][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:45,359][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:45,630][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:46,410][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:46,952][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:49,152][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:50,389][onir_pt][DEBUG] [finished] train pairs: [6.29s] [1024it] [162.88it/s]
[2024-04-20 21:09:50,391][onir_pt][DEBUG] [finished] training [6.29s]
[2024-04-20 21:09:50,392][onir_pt][INFO] training   it=5 loss=0.2221
[2024-04-20 21:09:50,392][onir_pt][DEBUG] [starting] validation
[2024-04-20 21:09:50,392][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:09:50,393][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2024-04-20 21:09:50,587][onir_pt][DEBUG] [finished] batches: s] [63it] [324.51it/s]
[2024-04-20 21:09:50,717][onir_pt][DEBUG] [finished] validation s]
[2024-04-20 21:09:50,718][onir_pt][INFO] validation it=5 map=0.0061 ndcg=0.0155 P_10=0.0780
[2024-04-20 21:09:50,719][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:09:50,720][onir_pt][DEBUG] [starting] training
[2024-04-20 21:09:50,720][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2024-04-20 21:09:51,274][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:51,790][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:54,184][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:55,016][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:55,654][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:56,233][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:56,571][onir_pt][DEBUG] [finished] train pairs: [5.85s] [1024it] [175.06it/s]
[2024-04-20 21:09:56,572][onir_pt][DEBUG] [finished] training [5.85s]
[2024-04-20 21:09:56,577][onir_pt][INFO] training   it=6 loss=0.2123
[2024-04-20 21:09:56,578][onir_pt][DEBUG] [starting] validation
[2024-04-20 21:09:56,579][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:09:56,580][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2024-04-20 21:09:56,767][onir_pt][DEBUG] [finished] batches: s] [63it] [338.32it/s]
[2024-04-20 21:09:56,887][onir_pt][DEBUG] [finished] validation s]
[2024-04-20 21:09:56,888][onir_pt][INFO] validation it=6 map=0.0060 ndcg=0.0155 P_10=0.0720
[2024-04-20 21:09:56,889][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:09:56,890][onir_pt][DEBUG] [starting] training
[2024-04-20 21:09:56,891][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2024-04-20 21:09:58,112][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:58,408][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:58,897][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:59,089][onir_pt][DEBUG] not enough negs
[2024-04-20 21:09:59,287][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:01,175][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:01,851][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:03,366][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:03,736][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:03,913][onir_pt][DEBUG] [finished] train pairs: [7.02s] [1024it] [145.84it/s]
[2024-04-20 21:10:03,916][onir_pt][DEBUG] [finished] training [7.03s]
[2024-04-20 21:10:03,917][onir_pt][INFO] training   it=7 loss=0.2152
[2024-04-20 21:10:03,918][onir_pt][DEBUG] [starting] validation
[2024-04-20 21:10:03,919][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:10:03,920][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2024-04-20 21:10:04,119][onir_pt][DEBUG] [finished] batches: s] [63it] [318.39it/s]
[2024-04-20 21:10:04,253][onir_pt][DEBUG] [finished] validation s]
[2024-04-20 21:10:04,255][onir_pt][INFO] validation it=7 map=0.0059 ndcg=0.0153 P_10=0.0720
[2024-04-20 21:10:04,256][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:10:04,258][onir_pt][DEBUG] [starting] training
[2024-04-20 21:10:04,258][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2024-04-20 21:10:04,349][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:05,274][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:05,637][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:05,729][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:06,297][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:07,596][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:08,304][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:08,581][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:08,845][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:09,192][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:09,980][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:10,403][onir_pt][DEBUG] [finished] train pairs: [6.14s] [1024it] [166.65it/s]
[2024-04-20 21:10:10,408][onir_pt][DEBUG] [finished] training [6.15s]
[2024-04-20 21:10:10,409][onir_pt][INFO] training   it=8 loss=0.2287
[2024-04-20 21:10:10,410][onir_pt][DEBUG] [starting] validation
[2024-04-20 21:10:10,410][onir_pt][DEBUG] using GPU (determinis

batches:   0%|          | 0/63 s<?, ?it/s]

[2024-04-20 21:10:10,595][onir_pt][DEBUG] [finished] batches: s] [63it] [345.26it/s]
[2024-04-20 21:10:10,739][onir_pt][DEBUG] [finished] validation s]
[2024-04-20 21:10:10,740][onir_pt][INFO] validation it=8 map=0.0060 ndcg=0.0153 P_10=0.0700
[2024-04-20 21:10:10,741][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:10:10,741][onir_pt][DEBUG] [starting] training
[2024-04-20 21:10:10,742][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2024-04-20 21:10:10,826][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:11,906][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:12,801][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:13,048][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:13,903][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:14,773][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:15,761][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:16,779][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:16,826][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:17,494][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:17,539][onir_pt][DEBUG] [finished] train pairs: [6.80s] [1024it] [150.68it/s]
[2024-04-20 21:10:17,543][onir_pt][DEBUG] [finished] training [6.80s]
[2024-04-20 21:10:17,544][onir_pt][INFO] training   it=9 loss=0.2266
[2024-04-20 21:10:17,544][onir_pt][DEBUG] [starting] validation
[2024-04-20 21:10:17,545][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:10:17,547][onir_pt][DEBUG] [starting] 

batches:   0%|          | 0/63 s<?, ?it/s]

[2024-04-20 21:10:17,733][onir_pt][DEBUG] [finished] batches: s] [63it] [340.13it/s]
[2024-04-20 21:10:17,853][onir_pt][DEBUG] [finished] validation s]
[2024-04-20 21:10:17,854][onir_pt][INFO] validation it=9 map=0.0060 ndcg=0.0153 P_10=0.0700
[2024-04-20 21:10:17,855][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:10:17,856][onir_pt][DEBUG] [starting] training
[2024-04-20 21:10:17,857][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2024-04-20 21:10:18,843][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:19,059][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:19,668][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:20,090][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:20,191][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:21,566][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:21,655][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:22,606][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:23,287][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:23,967][onir_pt][DEBUG] [finished] train pairs: [6.11s] [1024it] [167.60it/s]
[2024-04-20 21:10:23,969][onir_pt][DEBUG] [finished] training [6.11s]
[2024-04-20 21:10:23,974][onir_pt][INFO] training   it=10 loss=0.2253
[2024-04-20 21:10:23,975][onir_pt][DEBUG] [starting] validation
[2024-04-20 21:10:23,976][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:10:23,977][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2024-04-20 21:10:24,176][onir_pt][DEBUG] [finished] batches: s] [63it] [319.30it/s]
[2024-04-20 21:10:24,296][onir_pt][DEBUG] [finished] validation s]
[2024-04-20 21:10:24,297][onir_pt][INFO] validation it=10 map=0.0060 ndcg=0.0153 P_10=0.0700
[2024-04-20 21:10:24,298][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:10:24,299][onir_pt][DEBUG] [starting] training
[2024-04-20 21:10:24,300][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2024-04-20 21:10:25,250][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:25,370][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:25,825][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:26,522][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:27,692][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:28,930][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:29,841][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:30,896][onir_pt][DEBUG] [finished] train pairs: [6.59s] [1024it] [155.27it/s]
[2024-04-20 21:10:30,900][onir_pt][DEBUG] [finished] training [6.60s]
[2024-04-20 21:10:30,901][onir_pt][INFO] training   it=11 loss=0.2159
[2024-04-20 21:10:30,901][onir_pt][DEBUG] [starting] validation
[2024-04-20 21:10:30,902][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:10:30,906][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2024-04-20 21:10:31,231][onir_pt][DEBUG] [finished] batches: s] [63it] [194.52it/s]
[2024-04-20 21:10:31,489][onir_pt][DEBUG] [finished] validation s]
[2024-04-20 21:10:31,491][onir_pt][INFO] validation it=11 map=0.0060 ndcg=0.0153 P_10=0.0700
[2024-04-20 21:10:31,491][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:10:31,492][onir_pt][DEBUG] [starting] training
[2024-04-20 21:10:31,492][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2024-04-20 21:10:31,979][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:32,252][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:34,536][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:35,210][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:35,384][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:35,762][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:36,376][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:37,121][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:37,650][onir_pt][DEBUG] [finished] train pairs: [6.16s] [1024it] [166.34it/s]
[2024-04-20 21:10:37,651][onir_pt][DEBUG] [finished] training [6.16s]
[2024-04-20 21:10:37,654][onir_pt][INFO] training   it=12 loss=0.2180
[2024-04-20 21:10:37,655][onir_pt][DEBUG] [starting] validation
[2024-04-20 21:10:37,656][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:10:37,657][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2024-04-20 21:10:37,859][onir_pt][DEBUG] [finished] batches: s] [63it] [313.41it/s]
[2024-04-20 21:10:37,981][onir_pt][DEBUG] [finished] validation s]
[2024-04-20 21:10:37,982][onir_pt][INFO] validation it=12 map=0.0059 ndcg=0.0153 P_10=0.0700
[2024-04-20 21:10:37,983][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:10:37,984][onir_pt][DEBUG] [starting] training
[2024-04-20 21:10:37,984][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2024-04-20 21:10:38,421][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:38,521][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:39,694][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:40,023][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:42,438][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:43,005][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:43,163][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:43,715][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:44,493][onir_pt][DEBUG] [finished] train pairs: [6.51s] [1024it] [157.34it/s]
[2024-04-20 21:10:44,497][onir_pt][DEBUG] [finished] training [6.51s]
[2024-04-20 21:10:44,500][onir_pt][INFO] training   it=13 loss=0.2379
[2024-04-20 21:10:44,501][onir_pt][DEBUG] [starting] validation
[2024-04-20 21:10:44,501][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:10:44,502][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2024-04-20 21:10:44,781][onir_pt][DEBUG] [finished] batches: s] [63it] [226.87it/s]
[2024-04-20 21:10:45,057][onir_pt][DEBUG] [finished] validation s]
[2024-04-20 21:10:45,058][onir_pt][INFO] validation it=13 map=0.0054 ndcg=0.0143 P_10=0.0520
[2024-04-20 21:10:45,059][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:10:45,060][onir_pt][DEBUG] [starting] training
[2024-04-20 21:10:45,060][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2024-04-20 21:10:45,570][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:46,318][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:47,579][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:49,164][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:49,452][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:51,341][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:51,401][onir_pt][DEBUG] [finished] train pairs: [6.34s] [1024it] [161.53it/s]
[2024-04-20 21:10:51,403][onir_pt][DEBUG] [finished] training [6.34s]
[2024-04-20 21:10:51,403][onir_pt][INFO] training   it=14 loss=0.2261
[2024-04-20 21:10:51,403][onir_pt][DEBUG] [starting] validation
[2024-04-20 21:10:51,403][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:10:51,404][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2024-04-20 21:10:51,632][onir_pt][DEBUG] [finished] batches: s] [63it] [277.02it/s]
[2024-04-20 21:10:51,767][onir_pt][DEBUG] [finished] validation s]
[2024-04-20 21:10:51,768][onir_pt][INFO] validation it=14 map=0.0059 ndcg=0.0152 P_10=0.0700
[2024-04-20 21:10:51,769][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:10:51,769][onir_pt][DEBUG] [starting] training
[2024-04-20 21:10:51,770][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2024-04-20 21:10:51,812][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:52,172][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:53,176][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:53,519][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:53,829][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:55,464][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:56,855][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:57,017][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:57,320][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:57,740][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:58,064][onir_pt][DEBUG] [finished] train pairs: [6.29s] [1024it] [162.72it/s]
[2024-04-20 21:10:58,068][onir_pt][DEBUG] [finished] training [6.30s]
[2024-04-20 21:10:58,068][onir_pt][INFO] training   it=15 loss=0.2235
[2024-04-20 21:10:58,069][onir_pt][DEBUG] [starting] validation
[2024-04-20 21:10:58,070][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:10:58,071][onir_pt][DEBUG] [starting]

batches:   0%|          | 0/63 s<?, ?it/s]

[2024-04-20 21:10:58,369][onir_pt][DEBUG] [finished] batches: s] [63it] [212.12it/s]
[2024-04-20 21:10:58,611][onir_pt][DEBUG] [finished] validation s]
[2024-04-20 21:10:58,612][onir_pt][INFO] validation it=15 map=0.0060 ndcg=0.0154 P_10=0.0700
[2024-04-20 21:10:58,613][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:10:58,614][onir_pt][DEBUG] [starting] training
[2024-04-20 21:10:58,615][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2024-04-20 21:10:59,104][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:59,687][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:59,805][onir_pt][DEBUG] not enough negs
[2024-04-20 21:10:59,971][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:00,253][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:00,733][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:03,956][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:04,052][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:05,251][onir_pt][DEBUG] [finished] train pairs: [6.64s] [1024it] [154.32it/s]
[2024-04-20 21:11:05,256][onir_pt][DEBUG] [finished] training [6.64s]
[2024-04-20 21:11:05,259][onir_pt][INFO] training   it=16 loss=0.2301
[2024-04-20 21:11:05,260][onir_pt][DEBUG] [starting] validation
[2024-04-20 21:11:05,263][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:11:05,264][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2024-04-20 21:11:05,456][onir_pt][DEBUG] [finished] batches: s] [63it] [329.75it/s]
[2024-04-20 21:11:05,577][onir_pt][DEBUG] [finished] validation s]
[2024-04-20 21:11:05,578][onir_pt][INFO] validation it=16 map=0.0060 ndcg=0.0154 P_10=0.0700
[2024-04-20 21:11:05,579][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:11:05,580][onir_pt][DEBUG] [starting] training
[2024-04-20 21:11:05,580][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2024-04-20 21:11:05,828][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:06,795][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:07,079][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:07,756][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:07,921][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:08,666][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:08,843][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:10,018][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:10,318][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:10,442][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:11,106][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:12,101][onir_pt][DEBUG] [finished] train pairs: [6.52s] [1024it] [157.06it/s]
[2024-04-20 21:11:12,105][onir_pt][DEBUG] [finished] training [6.52s]
[2024-04-20 21:11:12,105][onir_pt][INFO] training   it=17 loss=0.2350
[2024-04-20 21:11:12,106][onir_pt][DEBUG] [starting] validation
[2024-04-20 21:11:12,107][onir_pt][DEBUG] using GPU (determini

batches:   0%|          | 0/63 s<?, ?it/s]

[2024-04-20 21:11:12,387][onir_pt][DEBUG] [finished] batches: s] [63it] [225.85it/s]
[2024-04-20 21:11:12,618][onir_pt][DEBUG] [finished] validation s]
[2024-04-20 21:11:12,619][onir_pt][INFO] validation it=17 map=0.0061 ndcg=0.0155 P_10=0.0720
[2024-04-20 21:11:12,620][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:11:12,621][onir_pt][DEBUG] [starting] training
[2024-04-20 21:11:12,621][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2024-04-20 21:11:12,673][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:12,796][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:13,308][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:13,603][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:14,272][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:15,561][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:16,125][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:17,043][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:18,922][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:19,466][onir_pt][DEBUG] [finished] train pairs: [6.84s] [1024it] [149.63it/s]
[2024-04-20 21:11:19,470][onir_pt][DEBUG] [finished] training [6.85s]
[2024-04-20 21:11:19,471][onir_pt][INFO] training   it=18 loss=0.2180
[2024-04-20 21:11:19,471][onir_pt][DEBUG] [starting] validation
[2024-04-20 21:11:19,472][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:11:19,473][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2024-04-20 21:11:19,706][onir_pt][DEBUG] [finished] batches: s] [63it] [271.49it/s]
[2024-04-20 21:11:19,831][onir_pt][DEBUG] [finished] validation s]
[2024-04-20 21:11:19,832][onir_pt][INFO] validation it=18 map=0.0059 ndcg=0.0152 P_10=0.0700
[2024-04-20 21:11:19,833][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:11:19,834][onir_pt][DEBUG] [starting] training
[2024-04-20 21:11:19,834][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2024-04-20 21:11:20,604][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:20,890][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:22,628][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:24,019][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:24,258][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:24,513][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:24,780][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:26,127][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:26,183][onir_pt][DEBUG] [finished] train pairs: [6.35s] [1024it] [161.32it/s]
[2024-04-20 21:11:26,187][onir_pt][DEBUG] [finished] training [6.35s]
[2024-04-20 21:11:26,187][onir_pt][INFO] training   it=19 loss=0.2323
[2024-04-20 21:11:26,188][onir_pt][DEBUG] [starting] validation
[2024-04-20 21:11:26,189][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:11:26,190][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2024-04-20 21:11:26,467][onir_pt][DEBUG] [finished] batches: s] [63it] [227.48it/s]
[2024-04-20 21:11:26,709][onir_pt][DEBUG] [finished] validation s]
[2024-04-20 21:11:26,711][onir_pt][INFO] validation it=19 map=0.0060 ndcg=0.0153 P_10=0.0700
[2024-04-20 21:11:26,711][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:11:26,712][onir_pt][DEBUG] [starting] training
[2024-04-20 21:11:26,713][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2024-04-20 21:11:27,210][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:27,289][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:27,480][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:27,731][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:29,802][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:30,549][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:31,137][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:32,138][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:33,364][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:33,564][onir_pt][DEBUG] [finished] train pairs: [6.85s] [1024it] [149.48it/s]
[2024-04-20 21:11:33,569][onir_pt][DEBUG] [finished] training [6.86s]
[2024-04-20 21:11:33,570][onir_pt][INFO] training   it=20 loss=0.2079
[2024-04-20 21:11:33,570][onir_pt][DEBUG] [starting] validation
[2024-04-20 21:11:33,571][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:11:33,572][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2024-04-20 21:11:33,804][onir_pt][DEBUG] [finished] batches: s] [63it] [272.76it/s]
[2024-04-20 21:11:33,925][onir_pt][DEBUG] [finished] validation s]
[2024-04-20 21:11:33,926][onir_pt][INFO] validation it=20 map=0.0060 ndcg=0.0154 P_10=0.0700
[2024-04-20 21:11:33,927][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:11:33,928][onir_pt][DEBUG] [starting] training
[2024-04-20 21:11:33,928][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2024-04-20 21:11:34,224][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:34,856][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:35,044][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:35,506][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:37,019][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:37,392][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:38,223][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:38,708][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:39,013][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:40,381][onir_pt][DEBUG] [finished] train pairs: [6.45s] [1024it] [158.70it/s]
[2024-04-20 21:11:40,386][onir_pt][DEBUG] [finished] training [6.46s]
[2024-04-20 21:11:40,387][onir_pt][INFO] training   it=21 loss=0.2284
[2024-04-20 21:11:40,388][onir_pt][DEBUG] [starting] validation
[2024-04-20 21:11:40,388][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:11:40,389][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2024-04-20 21:11:40,656][onir_pt][DEBUG] [finished] batches: s] [63it] [237.23it/s]
[2024-04-20 21:11:40,893][onir_pt][DEBUG] [finished] validation s]
[2024-04-20 21:11:40,894][onir_pt][INFO] validation it=21 map=0.0060 ndcg=0.0153 P_10=0.0700
[2024-04-20 21:11:40,894][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:11:40,895][onir_pt][DEBUG] [starting] training
[2024-04-20 21:11:40,896][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2024-04-20 21:11:41,163][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:41,787][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:42,089][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:42,722][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:43,964][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:45,090][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:45,564][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:45,976][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:45,983][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:47,467][onir_pt][DEBUG] [finished] train pairs: [6.57s] [1024it] [155.87it/s]
[2024-04-20 21:11:47,471][onir_pt][DEBUG] [finished] training [6.58s]
[2024-04-20 21:11:47,471][onir_pt][INFO] training   it=22 loss=0.2239
[2024-04-20 21:11:47,471][onir_pt][DEBUG] [starting] validation
[2024-04-20 21:11:47,471][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:11:47,472][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2024-04-20 21:11:47,769][onir_pt][DEBUG] [finished] batches: s] [63it] [211.96it/s]
[2024-04-20 21:11:48,032][onir_pt][DEBUG] [finished] validation s]
[2024-04-20 21:11:48,033][onir_pt][INFO] validation it=22 map=0.0060 ndcg=0.0154 P_10=0.0700
[2024-04-20 21:11:48,034][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:11:48,034][onir_pt][DEBUG] [starting] training
[2024-04-20 21:11:48,035][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2024-04-20 21:11:48,747][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:49,118][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:49,462][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:50,282][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:50,827][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:53,313][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:53,953][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:54,270][onir_pt][DEBUG] not enough negs
[2024-04-20 21:11:54,279][onir_pt][DEBUG] [finished] train pairs: [6.24s] [1024it] [164.02it/s]
[2024-04-20 21:11:54,282][onir_pt][DEBUG] [finished] training [6.25s]
[2024-04-20 21:11:54,283][onir_pt][INFO] training   it=23 loss=0.2235
[2024-04-20 21:11:54,283][onir_pt][DEBUG] [starting] validation
[2024-04-20 21:11:54,284][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:11:54,285][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2024-04-20 21:11:54,515][onir_pt][DEBUG] [finished] batches: s] [63it] [274.31it/s]
[2024-04-20 21:11:54,745][onir_pt][DEBUG] [finished] validation s]
[2024-04-20 21:11:54,746][onir_pt][INFO] validation it=23 map=0.0060 ndcg=0.0153 P_10=0.0700
[2024-04-20 21:11:54,747][onir_pt][INFO] early stopping; model reverting back to it=3


In [13]:
pt.Experiment(
    [tfidf, pipeline],
    test_topics,
    qrels,
    names=['TFIDF', 'TFIDF >> KNRM (trained)'],
    eval_metrics=[AP(rel=2), nDCG, nDCG@10, P(rel=2)@10]
)

[2024-04-20 21:11:59,407][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:11:59,408][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/188 s<?, ?it/s]

[2024-04-20 21:11:59,856][onir_pt][DEBUG] [finished] batches: s] [188it] [420.76it/s]


name  AP(rel=2)      nDCG   nDCG@10  P(rel=2)@10
0                    TFIDF   0.054568  0.123589  0.595818     0.546667
1  TFIDF >> KNRM (trained)   0.052371  0.121796  0.521478     0.453333

Le résultat est plus élévé, mais cela reste moins bon que le TFIDF. Proposez une hypothèse sur le problème.

le modèle KNRM entraîné n'ait pas donné de résultats aussi bons que le modèle TF-IDF seul, ce qui est contre-intuitif étant donné que l'objectif de l'entraînement du modèle est généralement d'améliorer les performances par rapport à une méthode de base comme TF-IDF. Voici quelques hypothèses possibles sur le problème :

1. **Surajustement (overfitting) du modèle KNRM** : le modèle KNRM ait surajusté les données d'entraînement, ce qui signifie qu'il a trop bien appris les exemples d'entraînement spécifiques mais qu'il ne généralise pas bien aux nouvelles données. Cela peut se produire si le modèle est trop complexe par rapport à la quantité de données d'entraînement disponibles.

2. **Paramètres d'entraînement non optimaux** : Les performances du modèle KNRM peuvent être sensibles aux paramètres d'entraînement tels que le taux d'apprentissage, le nombre d'époques d'entraînement, la taille du lot, etc. Il est possible que les paramètres que vous avez utilisés ne soient pas optimaux pour votre ensemble de données spécifique.

3. **Mauvaise initialisation des poids** : L'initialisation des poids du modèle KNRM peut jouer un rôle important dans ses performances. Si les poids sont initialisés de manière inappropriée, cela peut conduire à des difficultés lors de la convergence de l'entraînement ou à des performances médiocres sur les données de test.

4. **Déséquilibre des données d'entraînement** : Si les données d'entraînement sont déséquilibrées, c'est-à-dire si elles contiennent un nombre disproportionné de exemples positifs ou négatifs, cela peut affecter les performances du modèle KNRM, en particulier s'il n'est pas correctement traité.

5. **Erreur dans le prétraitement des données** : Il est également possible qu'il y ait eu une erreur dans le prétraitement des données avant l'entraînement du modèle KNRM, ce qui pourrait avoir un impact sur ses performances.



# 5. Vanilla BERT

Les modèles linguistiques contextualisés, tels que [BERT](https://arxiv.org/abs/1810.04805), sont des modèles neuronaux beaucoup plus puissants qui se sont avérés efficaces pour le classement.

Nous allons essayer d'utiliser une version "vanille" (ou "mono") du modèle BERT. Le modèle BERT est pré-entraîné pour la modélisation du langage et la prédiction de la phrase suivante.

In [14]:
#del knrm # clear out memory from KNRM
vbert = onir_pt.reranker('vanilla_transformer', 'bert', text_field='abstract', vocab_config={'train': True})

100%|██████████| 231508/231508 s<0ms, 1291708.75B/s]
100%|██████████| 433/433 s<0ms, 301133.08B/s]
100%|██████████| 440473133/440473133 [10.20s<0ms, 43203888.06B/s] 


Voyons comment ce modèle se comporte sur la collection TREC COVID.

In [15]:
pipeline = tfidf % 50 >> get_text >> vbert
pt.Experiment(
    [tfidf, pipeline],
    test_topics,
    qrels,
    names=['TFIDF', 'TFIDF >> VBERT'],
    baseline=0,
    eval_metrics=[AP(rel=2), nDCG, nDCG@10, P(rel=2)@10]
)

[2024-04-20 21:12:49,649][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:12:49,807][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/188 s<?, ?it/s]

[2024-04-20 21:13:11,183][onir_pt][DEBUG] [finished] batches: [21.37s] [188it] [ 8.80it/s]


name      nDCG   nDCG@10  AP(rel=2)  P(rel=2)@10  nDCG +  nDCG -  \
0           TFIDF  0.123589  0.595818   0.054568     0.546667     NaN     NaN   
1  TFIDF >> VBERT  0.118277  0.479929   0.049198     0.420000     3.0    12.0   

   nDCG p-value  nDCG@10 +  nDCG@10 -  nDCG@10 p-value  AP(rel=2) +  \
0           NaN        NaN        NaN              NaN          NaN   
1       0.11542        4.0       11.0          0.02287          3.0   

   AP(rel=2) -  AP(rel=2) p-value  P(rel=2)@10 +  P(rel=2)@10 -  \
0          NaN                NaN            NaN            NaN   
1         12.0           0.012239            3.0           12.0   

   P(rel=2)@10 p-value  
0                  NaN  
1             0.024362

In [16]:
pipeline.fit(
    train_topics,
    train_qrels,
    valid_topics,
    valid_qrels)

[2024-04-20 21:13:40,473][onir_pt][DEBUG] [starting] validation
[2024-04-20 21:13:40,474][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:13:40,478][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2024-04-20 21:13:47,636][onir_pt][DEBUG] [finished] batches: [7.16s] [63it] [ 8.80it/s]
[2024-04-20 21:13:47,759][onir_pt][DEBUG] [finished] validation [7.29s]
[2024-04-20 21:13:47,760][onir_pt][INFO] pre-validation: 0.0059
[2024-04-20 21:13:49,356][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:13:49,360][onir_pt][DEBUG] [starting] training
[2024-04-20 21:13:49,361][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2024-04-20 21:13:53,621][onir_pt][DEBUG] not enough negs
[2024-04-20 21:14:12,199][onir_pt][DEBUG] not enough negs
[2024-04-20 21:14:50,112][onir_pt][DEBUG] not enough negs
[2024-04-20 21:15:30,627][onir_pt][DEBUG] not enough negs
[2024-04-20 21:15:47,380][onir_pt][DEBUG] not enough negs
[2024-04-20 21:16:25,391][onir_pt][DEBUG] not enough negs
[2024-04-20 21:17:05,278][onir_pt][DEBUG] [finished] train pairs: [03:16] [1024it] [ 5.23it/s]
[2024-04-20 21:17:05,280][onir_pt][DEBUG] [finished] training [03:16]
[2024-04-20 21:17:05,284][onir_pt][INFO] training   it=0 loss=0.2504
[2024-04-20 21:17:05,287][onir_pt][DEBUG] [starting] validation
[2024-04-20 21:17:05,287][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:17:05,295][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2024-04-20 21:17:12,601][onir_pt][DEBUG] [finished] batches: [7.31s] [63it] [ 8.62it/s]
[2024-04-20 21:17:12,730][onir_pt][DEBUG] [finished] validation [7.44s]
[2024-04-20 21:17:12,731][onir_pt][INFO] validation it=0 map=0.0059 ndcg=0.0154 P_10=0.0660
[2024-04-20 21:17:12,732][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:17:12,736][onir_pt][DEBUG] [starting] training
[2024-04-20 21:17:12,736][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2024-04-20 21:17:21,742][onir_pt][DEBUG] not enough negs
[2024-04-20 21:17:25,090][onir_pt][DEBUG] not enough negs
[2024-04-20 21:18:32,800][onir_pt][DEBUG] not enough negs
[2024-04-20 21:19:09,036][onir_pt][DEBUG] not enough negs
[2024-04-20 21:19:19,677][onir_pt][DEBUG] not enough negs
[2024-04-20 21:19:35,991][onir_pt][DEBUG] not enough negs
[2024-04-20 21:19:40,510][onir_pt][DEBUG] not enough negs
[2024-04-20 21:19:59,684][onir_pt][DEBUG] not enough negs
[2024-04-20 21:20:20,206][onir_pt][DEBUG] not enough negs
[2024-04-20 21:20:27,201][onir_pt][DEBUG] not enough negs
[2024-04-20 21:20:31,933][onir_pt][DEBUG] [finished] train pairs: [03:19] [1024it] [ 5.14it/s]
[2024-04-20 21:20:31,935][onir_pt][DEBUG] [finished] training [03:19]
[2024-04-20 21:20:31,939][onir_pt][INFO] training   it=1 loss=0.2494
[2024-04-20 21:20:31,940][onir_pt][DEBUG] [starting] validation
[2024-04-20 21:20:31,940][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:20:31,955][onir_pt][DEBUG] [starting] b

batches:   0%|          | 0/63 s<?, ?it/s]

[2024-04-20 21:20:39,345][onir_pt][DEBUG] [finished] batches: [7.39s] [63it] [ 8.53it/s]
[2024-04-20 21:20:39,473][onir_pt][DEBUG] [finished] validation [7.53s]
[2024-04-20 21:20:46,464][onir_pt][INFO] validation it=1 map=0.0060 ndcg=0.0154 P_10=0.0720 <--
[2024-04-20 21:20:46,465][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:20:46,470][onir_pt][DEBUG] [starting] training
[2024-04-20 21:20:46,470][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2024-04-20 21:21:06,963][onir_pt][DEBUG] not enough negs
[2024-04-20 21:21:22,295][onir_pt][DEBUG] not enough negs
[2024-04-20 21:21:43,263][onir_pt][DEBUG] not enough negs
[2024-04-20 21:21:49,285][onir_pt][DEBUG] not enough negs
[2024-04-20 21:21:57,187][onir_pt][DEBUG] not enough negs
[2024-04-20 21:22:10,269][onir_pt][DEBUG] not enough negs
[2024-04-20 21:22:17,192][onir_pt][DEBUG] not enough negs
[2024-04-20 21:23:06,976][onir_pt][DEBUG] not enough negs
[2024-04-20 21:23:59,264][onir_pt][DEBUG] not enough negs
[2024-04-20 21:24:05,696][onir_pt][DEBUG] [finished] train pairs: [03:19] [1024it] [ 5.14it/s]
[2024-04-20 21:24:05,701][onir_pt][DEBUG] [finished] training [03:19]
[2024-04-20 21:24:05,705][onir_pt][INFO] training   it=2 loss=0.2478
[2024-04-20 21:24:05,707][onir_pt][DEBUG] [starting] validation
[2024-04-20 21:24:05,708][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:24:05,715][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2024-04-20 21:24:12,940][onir_pt][DEBUG] [finished] batches: [7.22s] [63it] [ 8.72it/s]
[2024-04-20 21:24:13,079][onir_pt][DEBUG] [finished] validation [7.37s]
[2024-04-20 21:24:15,070][onir_pt][INFO] validation it=2 map=0.0061 ndcg=0.0157 P_10=0.0760 <--
[2024-04-20 21:24:15,071][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:24:15,078][onir_pt][DEBUG] [starting] training
[2024-04-20 21:24:15,078][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2024-04-20 21:24:15,432][onir_pt][DEBUG] not enough negs
[2024-04-20 21:25:18,953][onir_pt][DEBUG] not enough negs
[2024-04-20 21:25:34,666][onir_pt][DEBUG] not enough negs
[2024-04-20 21:25:49,375][onir_pt][DEBUG] not enough negs
[2024-04-20 21:26:00,596][onir_pt][DEBUG] not enough negs
[2024-04-20 21:26:08,757][onir_pt][DEBUG] not enough negs
[2024-04-20 21:26:49,566][onir_pt][DEBUG] not enough negs
[2024-04-20 21:27:11,705][onir_pt][DEBUG] not enough negs
[2024-04-20 21:27:31,728][onir_pt][DEBUG] [finished] train pairs: [03:17] [1024it] [ 5.21it/s]
[2024-04-20 21:27:31,734][onir_pt][DEBUG] [finished] training [03:17]
[2024-04-20 21:27:31,736][onir_pt][INFO] training   it=3 loss=0.2574
[2024-04-20 21:27:31,738][onir_pt][DEBUG] [starting] validation
[2024-04-20 21:27:31,739][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:27:31,749][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2024-04-20 21:27:38,969][onir_pt][DEBUG] [finished] batches: [7.22s] [63it] [ 8.73it/s]
[2024-04-20 21:27:39,094][onir_pt][DEBUG] [finished] validation [7.36s]
[2024-04-20 21:27:39,095][onir_pt][INFO] validation it=3 map=0.0061 ndcg=0.0157 P_10=0.0740
[2024-04-20 21:27:39,096][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:27:39,100][onir_pt][DEBUG] [starting] training
[2024-04-20 21:27:39,101][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2024-04-20 21:27:51,329][onir_pt][DEBUG] not enough negs
[2024-04-20 21:28:05,269][onir_pt][DEBUG] not enough negs
[2024-04-20 21:28:12,279][onir_pt][DEBUG] not enough negs
[2024-04-20 21:28:51,180][onir_pt][DEBUG] not enough negs
[2024-04-20 21:29:12,688][onir_pt][DEBUG] not enough negs
[2024-04-20 21:29:27,172][onir_pt][DEBUG] not enough negs
[2024-04-20 21:30:32,584][onir_pt][DEBUG] not enough negs
[2024-04-20 21:30:38,080][onir_pt][DEBUG] not enough negs
[2024-04-20 21:30:54,415][onir_pt][DEBUG] [finished] train pairs: [03:15] [1024it] [ 5.24it/s]
[2024-04-20 21:30:54,417][onir_pt][DEBUG] [finished] training [03:15]
[2024-04-20 21:30:54,418][onir_pt][INFO] training   it=4 loss=0.2509
[2024-04-20 21:30:54,419][onir_pt][DEBUG] [starting] validation
[2024-04-20 21:30:54,419][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:30:54,435][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/63 s<?, ?it/s]

[2024-04-20 21:31:01,664][onir_pt][DEBUG] [finished] batches: [7.23s] [63it] [ 8.72it/s]
[2024-04-20 21:31:01,791][onir_pt][DEBUG] [finished] validation [7.37s]
[2024-04-20 21:31:01,793][onir_pt][INFO] validation it=4 map=0.0059 ndcg=0.0156 P_10=0.0700
[2024-04-20 21:31:01,793][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:31:01,798][onir_pt][DEBUG] [starting] training
[2024-04-20 21:31:01,799][onir_pt][DEBUG] [starting] train pairs


train pairs:   0%|          | 0/1024 s<?, ?it/s]

[2024-04-20 21:31:04,933][onir_pt][DEBUG] not enough negs


KeyboardInterrupt: 

In [17]:
pt.Experiment(
    [tfidf, pipeline],
    test_topics,
    qrels,
    names=['TFIDF', 'TFIDF >> VBERT'],
    eval_metrics=[AP(rel=2), nDCG, nDCG@10, P(rel=2)@10]
)

[2024-04-20 21:31:13,735][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:31:13,741][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/188 s<?, ?it/s]

[2024-04-20 21:31:35,311][onir_pt][DEBUG] [finished] batches: [21.57s] [188it] [ 8.72it/s]


name  AP(rel=2)      nDCG   nDCG@10  P(rel=2)@10
0           TFIDF   0.054568  0.123589  0.595818     0.546667
1  TFIDF >> VBERT   0.046396  0.116309  0.497435     0.413333

Comme nous le constatons, bien que le modèle soit pré-entraîné, il n'obtient pas de très bons résultats. Cela est dû au fait qu'il n'est pas réglé pour la tâche de classement par pertinence.

Cependant, nous pouvons entraîner le modèle pour le classement (comme indiqué ci-dessus pour KNRM).

# 6. monoT5

Le modèle [monoT5](https://arxiv.org/abs/2003.06713) évalue les documents à l'aide d'un modèle de langage causal. Voyons comment cette approche fonctionne sur TREC COVID.

La classe `MonoT5ReRanker` de `pyterrier_t5` charge automatiquement une version du classeur monoT5 qui est entraînée sur le jeu de données MS MARCO passage.

In [18]:
from pyterrier_t5 import MonoT5ReRanker
monoT5 = MonoT5ReRanker(text_field='abstract')

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:88: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


spiece.model:   0%|          | 0.00/792k s<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.39M s<?, ?B/s]

config.json:   0%|          | 0.00/1.21k s<?, ?B/s]

You are using the default legacy behaviour of the <class 'transformers.models.t5.tokenization_t5.T5Tokenizer'>. This is expected, and simply means that the `legacy` (previous) behavior will be used so nothing changes for you. If you want to use the new behaviour, set `legacy=False`. This should only be set if you understand what it means, and thoroughly read the reason why this was added as explained in https://github.com/huggingface/transformers/pull/24565
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.


config.json:   0%|          | 0.00/1.84k s<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/892M s<?, ?B/s]

In [19]:
pipeline = (tfidf >> get_text >> monoT5)
pt.Experiment(
    [tfidf, pipeline],
    test_topics,
    qrels,
    names=['TFIDF', 'TFIDF >> T5'],
    eval_metrics=[AP(rel=2), nDCG, nDCG@10, P(rel=2)@10, "mrt"]
)

monoT5:   0%|          | 0/188 s<?, ?batches/s]

name  AP(rel=2)      nDCG   nDCG@10  P(rel=2)@10          mrt
0        TFIDF   0.054568  0.123589  0.595818     0.546667    79.017215
1  TFIDF >> T5   0.058684  0.128119  0.685513     0.600000  1814.514595

Comme on pouvait s'y attendre, les résultats sont bien meilleurs en termes de NDCG@10 (0.5958 vs 0.6855).

# Tâche pratique

Comme pour le TP precedent, utilisez les modèles implémentés pour cord19 dans une tâche de question-réponse. Dans ce contexte, les requêtes sont de questions et les documents sont des documents qui pourraient contenir la réponse. Notez que vous devez refaire l'indexation ainsi que les autres étapes étudiées dans ce TP. Vous pouvez le télécharger le dataset en utilisant les lines de code ci-dessous.

In [20]:
fiqa = {}
fiqa['train'] = pt.datasets.get_dataset('irds:beir/fiqa/train')
fiqa['valid'] = pt.datasets.get_dataset('irds:beir/fiqa/dev')
fiqa['test'] = pt.datasets.get_dataset('irds:beir/fiqa/test')

test_topics = fiqa['test'].get_topics(variant='text')
test_qrels = fiqa['test'].get_qrels()

train_topics = fiqa['train'].get_topics(variant='text')
train_qrels = fiqa['train'].get_qrels()

valid_topics = fiqa['valid'].get_topics(variant='text')
valid_qrels = fiqa['valid'].get_qrels()

[INFO] [starting] opening zip file
[INFO] If you have a local copy of https://public.ukp.informatik.tu-darmstadt.de/thakur/BEIR/datasets/fiqa.zip, you can symlink it here to avoid downloading it again: /root/.ir_datasets/downloads/17918ed23cd04fb15047f73e6c3bd9d9
[INFO] [starting] https://public.ukp.informatik.tu-darmstadt.de/thakur/BEIR/datasets/fiqa.zip
[INFO] [finished] https://public.ukp.informatik.tu-darmstadt.de/thakur/BEIR/datasets/fiqa.zip: [1.91s] [17.9MB] [9.40MB/s]
[INFO] [finished] opening zip file [2.75s]
[INFO] [starting] opening zip file
[INFO] [finished] opening zip file s]
[INFO] [starting] opening zip file
[INFO] [finished] opening zip file s]
[INFO] [starting] opening zip file
[INFO] [finished] opening zip file s]


In [21]:
# réponse

import os
from itertools import chain

pt_index_path = './terrier_fiqa'

if not os.path.exists(pt_index_path + "/data.properties"):
    indexer = pt.index.IterDictIndexer(pt_index_path, blocks=True)

    index_ref = indexer.index(chain(fiqa['train'].get_corpus_iter(),
                                    fiqa['valid'].get_corpus_iter(),
                                    fiqa['test'].get_corpus_iter()),
                              fields=('text',),
                              meta=('docno',))

else:
    index_ref = pt.IndexRef.of(pt_index_path + "/data.properties")

[INFO] [starting] building docstore
[INFO] [starting] opening zip file
[INFO] [finished] opening zip file s]
docs_iter: 100%|████████████████████████| 57638/57638 [1.69s<0ms, 34056.67doc/s]
[INFO] [finished] docs_iter: [1.69s] [57638doc] [34034.81doc/s]
[INFO] [finished] building docstore [1.70s]


beir/fiqa/train documents:   0%|          | 0/57638 s<?, ?it/s]

beir/fiqa/dev documents:   0%|          | 0/57638 s<?, ?it/s]

beir/fiqa/test documents:   0%|          | 0/57638 s<?, ?it/s]

<ipython-input-21-c2bdfa48e16a>:11: DeprecationWarning: specifying meta and meta_lengths in IterDictIndexer.index() is deprecated, use constructor instead
  index_ref = indexer.index(chain(fiqa['train'].get_corpus_iter(),


21:32:36.791 [ForkJoinPool-2-worker-3] WARN org.terrier.structures.indexing.Indexer - Adding an empty document to the index (7915) - further warnings are suppressed
21:33:24.418 [ForkJoinPool-2-worker-3] ERROR org.terrier.structures.indexing.Indexer - Could not finish MetaIndexBuilder: 
java.io.IOException: Key 100 is not unique: 115281,57643
For MetaIndex, to suppress, set metaindex.compressed.reverse.allow.duplicates=true
	at org.terrier.structures.collections.FSOrderedMapFile$MultiFSOMapWriter.mergeTwo(FSOrderedMapFile.java:1374)
	at org.terrier.structures.collections.FSOrderedMapFile$MultiFSOMapWriter.close(FSOrderedMapFile.java:1308)
	at org.terrier.structures.indexing.BaseMetaIndexBuilder.close(BaseMetaIndexBuilder.java:321)
	at org.terrier.structures.indexing.classical.BasicIndexer.indexDocuments(BasicIndexer.java:270)
	at org.terrier.structures.indexing.classical.BasicIndexer.createDirectIndex(BasicIndexer.java:388)
	at org.terrier.structures.indexing.Indexer.index(Indexer.java

In [22]:
knrm = onir_pt.reranker('knrm', 'wordvec_hash', text_field='abstract',config={'max_train_it': 10,'random_seed': 42})


[2024-04-20 21:33:36,189][WordvecHashVocab][DEBUG] [starting] reading cached at /root/data/onir/vocab/wordvec_hash/fasttext-wiki-news-300d-1M.p
[2024-04-20 21:33:46,792][WordvecHashVocab][DEBUG] [finished] reading cached at /root/data/onir/vocab/wordvec_hash/fasttext-wiki-news-300d-1M.p [10.60s]


In [ ]:
pipeline = tfidf % 50 >> get_text >> vbert
pt.Experiment(
    [tfidf, pipeline],
    test_topics,
    qrels,

    baseline=0,
    eval_metrics=[AP(rel=2), nDCG, nDCG@10, P(rel=2)@10]
)

[2024-04-20 21:34:45,856][onir_pt][DEBUG] using GPU (deterministic)
[2024-04-20 21:34:45,872][onir_pt][DEBUG] [starting] batches


batches:   0%|          | 0/8100 s<?, ?it/s]

In [ ]:
pipeline.fit(
    train_topics,
    train_qrels,
    valid_topics,
    valid_qrels)

In [ ]:
pt.Experiment(
    [tfidf, pipeline],
    test_topics,
    qrels,
    names=['TFIDF', 'TFIDF >> KNRM (trained)'],
    eval_metrics=[AP(rel=2), nDCG, nDCG@10, P(rel=2)@10]
)

** Question**. Identifiez-vous des avantages par rapport à l'utilisation de Learning to Rank ?

---



**réponse**: Amelioration des resultats de recherche et des reponses parce que la base de donnee est une base de donnee de question reponse